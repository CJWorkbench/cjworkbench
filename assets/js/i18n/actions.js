const SET_LOCALE = 'SET_LOCALE';

function reduceSetLocale(state, action){
  return {
    ...state,
    locale: action.locale
  };
}

export function setLocaleAction(locale){
  return {
    type: SET_LOCALE,
    locale: locale
  }
}

export const reducerFunctions = {
  [SET_LOCALE]: reduceSetLocale
}

export function localeReducer(state, action){
  if (action.type in reducerFunctions) {
    // Run a registered reducer
    return reducerFunctions[action.type](state, action)
  }
  
  return state;
}
