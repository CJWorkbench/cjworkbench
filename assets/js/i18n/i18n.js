import { setupI18n } from "@lingui/core"
import { currentLocale } from './locales'
import { fetchCatalog } from './catalogs'

export default const i18n = setupI18n({
   language: currentLocale,
   catalogs: {
       currentLocale: fetchCatalog(currentLocale)
   }
})
