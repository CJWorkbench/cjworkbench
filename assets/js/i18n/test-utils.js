import { I18nProvider } from '@lingui/react'
import React from 'react'
import { shape, object } from 'prop-types'
import { mount, shallow } from 'enzyme'

/*
 * Below lie utils for testing with i18n.
 * See https://lingui.js.org/guides/testing.html
 */
// Create the I18nProvider to retrieve context for wrapping around.
const intlProvider = new I18nProvider({
  language: 'en',
  catalogs: {
    en: {}
  }
}, {})

const {
  linguiPublisher: {
    i18n: originalI18n
  }
} = intlProvider.getChildContext()

// You customize the i18n object here:
export const i18n = {
  ...originalI18n,
  // provide _ macro, for trying to unwrap default/id or just passing down the key
  _: key => {
    if (typeof key.defaults === 'string') return key.defaults
    if (typeof key.id === 'string') return key.id
    return key
  }
}

/**
 * When using Lingui `withI18n` on components, props.i18n is required.
 */
function nodeWithI18nProp (node) {
  return React.cloneElement(node, { i18n })
}

/*
 * Methods to use
 */
export function shallowWithI18n (node, { context } = {}) {
  return shallow(
    nodeWithI18nProp(node),
    {
      context: Object.assign({}, context, { i18n })
    }
  )
}

export function mountWithI18n (node, { context, childContextTypes } = {}) {
  const newContext = Object.assign({}, context, { linguiPublisher: { i18n } })
  /*
   * I18nProvider sets the linguiPublisher in the context for withI18n to get
   * the i18n object from.
   */
  const newChildContextTypes = Object.assign({},
    {
      linguiPublisher: shape({
        i18n: object.isRequired
      }).isRequired
    },
    childContextTypes
  )
  return mount(
    nodeWithI18nProp(node),
    {
      context: newContext,
      childContextTypes: newChildContextTypes
    }
  )
}
