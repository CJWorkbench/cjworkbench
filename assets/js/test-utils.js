/* globals jest */
import { createStore, applyMiddleware } from 'redux'
import thunk from 'redux-thunk'
import promiseMiddleware from 'redux-promise-middleware'
import errorMiddleware from './error-middleware'
import { workflowReducer } from './workflow-reducer'

// Returns new mock function that returns given json. Used for mocking "get" API calls
export function jsonResponseMock (json) {
  return jest.fn().mockImplementation(() =>
    Promise.resolve(json)
  )
}

// Returns new mock function that gives an OK HTTP response. Use for mocking "set" API calls
export function okResponseMock () {
  return jsonResponseMock(null)
}

// Helper function to return a promise that resolves after all other promise mocks,
// even if they are chained like Promise.resolve().then(...).then(...)
// Technically: this is designed to resolve on the next macrotask
// https://stackoverflow.com/questions/25915634/difference-between-microtask-and-macrotask-within-an-event-loop-context
export function tick () {
  return new Promise(resolve => {
    setTimeout(resolve, 0)
  })
}

export function sleep (ms) {
  return new Promise(resolve => {
    setTimeout(resolve, ms)
  })
}

/**
 * Return `[ notify, promise ]`; `promise` completes when `notify` is called.
 *
 * Usage:
 *
 *     const [ notify, condition ] = createConditionVariable
 *     const done = condition.then(do_something)
 *     await Promise.all([sleep(0.1).then(notify), done])
 */
export function createConditionVariable () {
  let notify
  const promise = new Promise(resolve => { notify = resolve })
  return [notify, promise]
}

/**
 * Mock Redux store with in-production reducers.
 *
 * This is for integration-style tests: they test the reducer _and_ the caller
 * at the same time.
 *
 * Usage:
 *
 *     const api = { undo: jest.fn() }
 *     const store = mockStoreWithReducer({'workflow': ..., ...}, api)
 *     store.dispatch(Actions.undo())
 *     expect(api.undo).toHaveBeenCalled()
 *     expect(store.getState().workflow).toEqual(...)
 */
export function mockStore (initialState, api = {}) {
  const middlewares = [errorMiddleware(), promiseMiddleware, thunk.withExtraArgument(api)]
  const store = createStore(workflowReducer, initialState, applyMiddleware(...middlewares))
  return store
}
