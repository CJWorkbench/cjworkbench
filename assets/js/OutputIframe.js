import React from 'react'
import PropTypes from 'prop-types'
import { Modal, ModalHeader, ModalBody, ModalFooter } from './components/Modal'
import { setWfModuleParamsAction } from './workflow-reducer'
import { setWorkflowPublicAction } from './ShareModal/actions'
import { connect } from 'react-redux'
import { escapeHtml } from './utils'
import { Trans, t } from '@lingui/macro'
import { withI18n } from '@lingui/react'

export class OutputIframe extends React.PureComponent {
  static propTypes = {
    visible: PropTypes.bool.isRequired, // false means, "zero height"
    deltaId: PropTypes.number, // null if added to empty workflow
    wfModuleId: PropTypes.number, // null if no wfmodule
    isPublic: PropTypes.bool.isRequired,
    workflowId: PropTypes.number.isRequired
  }

  state = {
    heightFromIframe: null, // if set, the iframe told us how tall it wants to be
    isModalOpen: false
  }

  componentDidUpdate (prevProps, prevState) {
    if (prevProps.wfModuleId !== this.props.wfModuleId) {
      this.setState({
        heightFromIframe: null
      })
    }

    if (prevState.heightFromIframe !== this.state.heightFromIframe || prevProps.visible !== this.props.visible) {
      const resizeEvent = document.createEvent('Event')
      resizeEvent.initEvent('resize', true, true)
      window.dispatchEvent(resizeEvent)
    }
  }

  componentDidMount () {
    window.addEventListener('message', this.handleMessage)
  }

  componentWillUnmount () {
    window.removeEventListener('message', this.handleMessage)
  }

  handleMessage = (ev) => {
    const data = ev.data
    if (data && data.from === 'outputIframe') {
      if (data.wfModuleId !== this.props.wfModuleId) {
        // This message isn't from the iframe we created.
        //
        // This check works around a race:
        //
        // 1. Show an iframe
        //    ... it sends a 'resize' event
        //    ... it keeps sending 'resize' events whenever its size changes
        // 2. Switch to a different iframe src
        //    ... this resets size and sets new iframe src; BUT before the new
        //        iframe can load, the _old_ iframe's JS sends a 'resize' event
        //
        // By forcing the iframe to send its identity, we can make sure this
        // message isn't spurious.
        return
      }

      switch (data.type) {
        case 'resize':
          this.setState({ heightFromIframe: data.height })
          break
        case 'set-params':
          this.props.setWfModuleParams(data.wfModuleId, data.params)
          break
        default:
          console.error('Unhandled message from iframe', data)
      }
    }
  }

  handleClickOpenEmbedModal = () => {
    this.setState({ isModalOpen: true })
  }

  closeModal = () => {
    this.setState({ isModalOpen: false })
  }

  handleClickModalClose = this.closeModal

  handleClickSetWorkflowPublic = () => { this.props.setWorkflowPublic() }

  isModalOpen (name) {
    if (!this.state.isModalOpen) return false
    if (this.props.isPublic) {
      return name === 'embed'
    } else {
      return name === 'public'
    }
  }

  renderPublicModal () {
    return (
      <Modal isOpen={this.isModalOpen('public')} toggle={this.closeModal}>
        <ModalHeader toggle={this.closeModal}>
          <div className='modal-title'>
            <Trans id='js.OutputIframe.private.header.title' description='This should be all-caps for styling reasons'>
                SHARE THIS WORKFLOW
            </Trans>
          </div>
        </ModalHeader>
        <ModalBody>
          <div className='title-3 mb-3'>
            <Trans id='js.OutputIframe.private.workflowIsPrivate'>This workflow is currently private</Trans>
          </div>
          <div className='info-3 t-d-gray'>
            <Trans id='js.OutputIframe.private.setToPublic'>
                Set this workflow to Public in order to share it? Anyone with the URL will be able to access and duplicate it.
            </Trans>
          </div>
        </ModalBody>
        <ModalFooter>
          <div onClick={this.handleClickModalClose} className='button-gray action-button mr-4'>
            <Trans id='js.OutputIframe.footer.cancelButton'>Cancel</Trans>
          </div>
          <div onClick={this.handleClickSetWorkflowPublic} className='button-blue action-button test-public-button'>
            <Trans id='js.OutputIframe.footer.setPublicButton'>Set public</Trans>
          </div>
        </ModalFooter>
      </Modal>
    )
  }

  renderEmbedModal () {
    const iframeCode = escapeHtml('<iframe src="' + window.location.protocol + '//' + window.location.host + '/embed/' + this.props.wfModuleId + '" width="560" height="315" frameborder="0" />')

    return (
      <Modal isOpen={this.isModalOpen('embed')} toggle={this.closeModal}>
        <ModalHeader toggle={this.closeModal}>
          <div className='modal-title'>
            <Trans id='js.OutputIframe.embed.header.title' description='This should be all-caps for styling reasons'>
                EMBED THIS CHART
            </Trans>
          </div>
        </ModalHeader>
        <ModalBody>
          <p className='info'>
            <Trans id='js.OutputIframe.embed.embedCode'>Paste this code into any webpage HTML</Trans>
          </p>
          <div className='code-snippet'>
            <code className='chart-embed'>
              {iframeCode}
            </code>
          </div>
        </ModalBody>
        <div className='modal-footer'>
          <div onClick={this.handleClickModalClose} className='button-gray action-button'>
            <Trans id='js.OutputIframe.footer.OKButton'>OK</Trans>
          </div>
        </div>
      </Modal>
    )
  }

  render () {
    const { wfModuleId, deltaId, visible, i18n } = this.props
    const { heightFromIframe } = this.state
    const src = `/api/wfmodules/${wfModuleId}/output#revision=${deltaId}`

    const defaultHeight = visible ? '100%' : '0'
    const height = heightFromIframe === null ? defaultHeight : `${Math.ceil(heightFromIframe)}px`

    return (
      <div className='outputpane-iframe' style={{ height }}>
        {!visible ? null : (
          <>
            <iframe src={src} />
            <div className='outputpane-iframe-control-overlay'>
              <button name='embed' className='btn' title={i18n._(t('js.OutputIframe.getEmbeddableUrl.hoverText')`Get an embeddable URL`)} onClick={this.handleClickOpenEmbedModal}>
                <i className='icon icon-code' />
              </button>
            </div>
            {this.renderPublicModal()}
            {this.renderEmbedModal()}
          </>
        )}
      </div>
    )
  }
}

const mapStateToProps = (state) => ({})

const mapDispatchToProps = (dispatch, ownProps) => {
  return {
    setWorkflowPublic: () => {
      dispatch(setWorkflowPublicAction(true))
    },
    setWfModuleParams: (wfModuleId, params) => {
      dispatch(setWfModuleParamsAction(wfModuleId, params))
    }
  }
}

export default connect(
  mapStateToProps,
  mapDispatchToProps
)(withI18n()(OutputIframe))
