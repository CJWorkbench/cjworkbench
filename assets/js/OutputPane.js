// Display of output from currently selected module

import React from 'react'
import TableView from './TableView'
import PropTypes from 'prop-types'
import { OutputIframe } from './OutputIframe'
import Resizable from 're-resizable'
import debounce from 'lodash/debounce'

export default class OutputPane extends React.Component {

  constructor(props) {
    super(props);

    this.state = {
        leftOffset : 0,
        initLeftOffset: 0,
        width: "100%",
        height: "100%",
        maxWidth: "300%",
        parentBase: null,
        pctBase: null,
        resizing: false
    };

    this.setBusySpinner = this.setBusySpinner.bind(this);
    this.saveSpinnerEl = this.saveSpinnerEl.bind(this);
    this.resizePaneStart = this.resizePaneStart.bind(this);
    this.resizePane = this.resizePane.bind(this);
    this.resizePaneEnd = this.resizePaneEnd.bind(this);
    this.setResizePaneRelativeDimensions = this.setResizePaneRelativeDimensions.bind(this);

    this.spinnerEl = null;
    this.spinning = false;
  }

  // Spinner state did not work as part of component state, conditionally visible in render()
  // It didn't appear when refreshing a large table. My guess is that is because React updates are batched,
  // the spinner on and spinner off updates are combined and we never see it when the table re-render is long.
  // So, now we turn the spinner on and off immediately through direct DOM styling
  setBusySpinner(visible) {
    if (this.spinnerEl && visible != this.spinning) {
      this.spinnerEl.style.display = visible ? 'flex' : 'none';
      this.spinning = visible;
    }
  }

  // Can't do this as an anonymous function like ref={ (el) => {this.spinnerEl=el} }
  // because el will sometimes be null if we do. See https://reactjs.org/docs/refs-and-the-dom.html#caveats
  saveSpinnerEl(el) {
    this.spinnerEl = el;
  }

  componentDidMount() {
    window.addEventListener("resize", debounce(() => { this.setResizePaneRelativeDimensions(this.props.libraryOpen) }, 200));
    this.setResizePaneRelativeDimensions(this.props.libraryOpen);
  }

  componentWillReceiveProps(nextProps) {
      console.log("OutputPane props");
      console.log(nextProps);
    if (nextProps.libraryOpen !== this.props.libraryOpen) {
        this.setResizePaneRelativeDimensions(nextProps.libraryOpen, true);
    }
  }

  getWindowWidth() {
      return window.innerWidth
        || document.documentElement.clientWidth
        || document.body.clientWidth;
  }

  resizePaneStart() {
      this.props.setOverlapping(true);
      this.props.setFocus();
  }

  resizePane(e, direction, ref, d) {
    let offset = this.state.initLeftOffset - d.width;
    this.setState({
        leftOffset: offset,
        resizing: true
    });
  }

  resizePaneEnd(e, direction, ref, d) {
      // We set percentage width so that we can maintain the outer layout of the module stack and output pane
      // while still allowing the right pane to stretch beyond its boundaries. Setting a pixel width on the inner
      // part of the output pane expands the outer frame, distorting the layout of the entire page.
      let width = parseFloat(this.state.width) + ((d.width / this.state.pctBase) * 100) + '%';
      this.setState({
          initLeftOffset: this.state.leftOffset,
          width,
          resizing: false
      });
      this.props.setOverlapping((this.state.leftOffset < 0));
  }

  /* Set the width and left offset of the resize pane relative to the window size and collapsed state of the
        module library. Deals with the following cases:

  1. Window resize -- re-position and resize right pane to new relative position with new maximum width relative to
        window size while maintaining the same visual offset from the left edge

  2. Open/close module library while right pane is at "0" -- re-position and resize right pane to "0" position relative
        to ML state

  3. Open/close module library while right pane is expanded but less than max: re-position and resize right pane so
        it maintains the same visual position on the screen

  4. Open module library while right pane is at max width relative to closed ML: re-position and re-size right pane to
        max width relative to open ML position
   */

  setResizePaneRelativeDimensions(libraryState, libraryToggle) {
      let libraryOffset = 0;
      let resetWidth;
      let resetOffset;
      let maxWidthOffset = 0;
      let resetMaxWidth;

      if (libraryState === true) {
          maxWidthOffset = 240;
          if (libraryToggle === true) {
              libraryOffset = -140;
          }
      }

      if (libraryState === false) {
          maxWidthOffset = 100;
          if (libraryToggle === true) {
              libraryOffset = 140;
          }

      }

      resetOffset = this.state.leftOffset + libraryOffset;

      if (resetOffset > 0 || this.state.leftOffset === 0) {
          resetOffset = 0;
          resetWidth = '100%';
      } else {
          resetWidth = ((this.state.parentBase.clientWidth - resetOffset) / this.state.parentBase.clientWidth) * 100 + '%';
      }

      resetMaxWidth = ((this.getWindowWidth() - maxWidthOffset) / this.state.parentBase.clientWidth) * 100 + '%';

      if ( parseFloat(resetWidth) > parseFloat(resetMaxWidth) ) {
          resetOffset = resetOffset + ( this.state.parentBase.clientWidth * ( ( parseFloat(resetWidth) - parseFloat(resetMaxWidth) ) / 100 ) );
          resetWidth = resetMaxWidth;
      }

      this.setState({
          leftOffset : resetOffset,
          initLeftOffset: resetOffset,
          width: resetWidth,
          height: "100%",
          maxWidth: resetMaxWidth,
          pctBase: this.state.parentBase.clientWidth
      });
  }

  findCurrentModuleInWorkflow(wf) {
      console.log(wf);
      var modulesFound = wf.wf_modules.filter((wfm) => {return wfm.id == this.props.selectedWfModuleId});
      console.log(modulesFound);
      return modulesFound.length > 0 ? modulesFound[0] : null;
  }

  render() {
    // We figure out whether we need to indicate sort status here so that we don't have to
    // pass a ton of data to the TableView

    var moduleIsSort = false;
    let currentModule = this.findCurrentModuleInWorkflow(this.props.workflow);
    console.log("OutputPane data:")
      console.log(this.props.workflow);
    console.log(currentModule)
    if(currentModule) {
        moduleIsSort = (currentModule.module_version.module.id_name == "sort-from-table")
    }
    console.log(moduleIsSort)

    // Maps sort direction to ReactDataGrid direction names
    let sortDirectionTranslator = ["NONE", "ASC", "DESC"]

    // Make a table component even if no module ID (should still show an empty table)
    var tableView =
      <TableView
        id={this.props.id}
        revision={this.props.revision}
        resizing={this.state.resizing}
        api={this.props.api}
        setBusySpinner={this.setBusySpinner}
        sortColumn={moduleIsSort ? currentModule.parameter_vals[0].value : undefined}
        sortDirection={moduleIsSort ? sortDirectionTranslator[currentModule.parameter_vals[2].value] : undefined}
      />

    // This iframe holds the module HTML output, e.g. a visualization
    var outputIFrame = null;
    if (this.props.htmlOutput) {
      outputIFrame = <OutputIframe
          api={this.props.api}
          selectedWfModuleId={this.props.selectedWfModuleId}
          workflow={this.props.workflow}
          revision={this.props.revision}
      />
    }

    // Spinner is always rendered, but we toggle 'display: none' in setBusySpinner()
    // Start hidden. TableView will turn it on when needed.
    var spinner =
      <div
        id="spinner-container-transparent"
        style={{display:'none'}}
        ref={ this.saveSpinnerEl }
      >
        <div id="spinner-l1">
          <div id="spinner-l2">
            <div id="spinner-l3"></div>
          </div>
        </div>
      </div>

    return (
        <div className={"outputpane" + (this.props.focus ? " focus" : "")}
             ref={(ref) => this.state.parentBase = ref}
             onClick={this.props.setFocus} >
            <Resizable
              style={{
                  transform: "translateX(" + this.state.leftOffset + "px)"
              }}
              className="outputpane-box"
              enable={{
                  top:false,
                  right:false,
                  bottom:false,
                  left:true,
                  topRight:false,
                  bottomRight:false,
                  bottomLeft:false,
                  topLeft:false
              }}
              size={{
                  width: this.state.width,
                  height: this.state.height,
              }}
              minWidth="100%"
              maxWidth={this.state.maxWidth}
              onResizeStart={this.resizePaneStart}
              onResize={this.resizePane}
              onResizeStop={this.resizePaneEnd}
            >

              {spinner}

              { outputIFrame }

              { tableView }

            </Resizable>
        </div>
    );
  }
}

OutputPane.propTypes = {
  id:                 PropTypes.number,             // can be undefined, if no selected module
  revision:           PropTypes.number.isRequired,
  api:                PropTypes.object.isRequired,
  selectedWfModuleId: PropTypes.number,
  htmlOutput:         PropTypes.bool,
};
