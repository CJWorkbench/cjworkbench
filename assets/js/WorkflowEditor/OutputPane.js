// Display of output from currently selected module

import React from 'react'
import PropTypes from 'prop-types'
import DelayedTableSwitcher from '../table/DelayedTableSwitcher'
import OutputIframe from '../OutputIframe'
import { connect } from 'react-redux'
import { withI18n } from '@lingui/react'
import { t } from '@lingui/macro'

export class OutputPane extends React.Component {
  static propTypes = {
    loadRows: PropTypes.func.isRequired, // func(wfModuleId, deltaId, startRowInclusive, endRowExclusive) => Promise[Array[Object] or error]
    workflowId: PropTypes.number.isRequired,
    wfModuleBeforeError: PropTypes.shape({
      id: PropTypes.number.isRequired,
      deltaId: PropTypes.number, // or null -- it may not be rendered
      status: PropTypes.oneOf(['ok', 'busy', 'unreachable']).isRequired, // can't be 'error'
      columns: PropTypes.arrayOf(PropTypes.shape({
        name: PropTypes.string.isRequired,
        type: PropTypes.oneOf(['text', 'number', 'datetime']).isRequired
      }).isRequired), // or null
      nRows: PropTypes.number // or null
    }), // or null if no error
    wfModule: PropTypes.shape({
      id: PropTypes.number.isRequired,
      htmlOutput: PropTypes.bool.isRequired,
      status: PropTypes.oneOf(['ok', 'busy', 'error', 'unreachable']).isRequired,
      deltaId: PropTypes.number, // or null if not yet rendered
      columns: PropTypes.arrayOf(PropTypes.shape({
        name: PropTypes.string.isRequired,
        type: PropTypes.oneOf(['text', 'number', 'datetime']).isRequired
      }).isRequired), // or null
      nRows: PropTypes.number // or null
    }), // or null if no selection
    isPublic: PropTypes.bool.isRequired,
    isReadOnly: PropTypes.bool.isRequired,
    i18n: PropTypes.shape({
      // i18n object injected by LinguiJS withI18n()
      _: PropTypes.func.isRequired
    })
  }

  /**
   * Return the WfModule we want to render for the user.
   *
   * This will _never_ be an "error"-status WfModule. If there's an error, we
   * want the user to see the input. (This component will also render a notice
   * saying it's showing the input.)
   */
  get wfModuleForTable () {
    const { wfModuleBeforeError, wfModule } = this.props

    if (wfModuleBeforeError) {
      // We're focused on an error module. The user wants to see its _input_ to
      // debug it.
      return wfModuleBeforeError
    } else if (wfModule && wfModule.status !== 'error') {
      return wfModule
    } else {
      // Either there's no selected WfModule, or the selected WfModule has
      // status === 'error' and it's the first in the tab. Either way, we want
      // to render a "placeholder" table.
      return null
    }
  }

  renderOutputIFrame () {
    // Always show _this_ module's iframe. If this module has status 'error'
    // and it's the Python console, the iframe contains the stack trace. If
    // we showed the _input_ module's iframe we wouldn't render the stack
    // trace.

    const { wfModule, workflowId, isPublic } = this.props

    const wfModuleId = wfModule ? wfModule.id : null
    const deltaId = wfModule ? wfModule.deltaId : null
    const htmlOutput = wfModule ? wfModule.htmlOutput : false

    // This iframe holds the module HTML output, e.g. a visualization.
    // We leave the component around even when there is no HTML because of
    // our solution to https://www.pivotaltracker.com/story/show/159637930:
    // DataGrid.js doesn't notice the resize that occurs when the iframe
    // appears or disappears.
    return (
      <OutputIframe
        key='iframe'
        visible={htmlOutput}
        workflowId={workflowId}
        isPublic={isPublic}
        wfModuleId={wfModuleId}
        deltaId={deltaId}
      />
    )
  }

  renderShowingInput () {
    if (this.props.wfModuleBeforeError) {
      return (
        <p
          key='error'
          className='showing-input-because-error'
        >
          {this.props.i18n._(t('js.WorkflowEditor.OutputPane.showingInput.becauseError')`This was the data that led to an error. Please correct the error in the left pane.`)}
        </p>
      )
    } else {
      return null
    }
  }

  render () {
    const { isReadOnly, loadRows, wfModule } = this.props
    const wfm = this.wfModuleForTable
    const className = 'outputpane module-' + (wfModule ? wfModule.status : 'unreachable')

    return (
      <div className={className}>
        {this.renderOutputIFrame()}
        {this.renderShowingInput()}
        <DelayedTableSwitcher
          key='table'
          wfModuleId={wfm ? wfm.id : null}
          status={wfm ? wfm.status : null}
          deltaId={wfm ? wfm.deltaId : null}
          columns={wfm ? wfm.columns : null}
          nRows={wfm ? wfm.nRows : null}
          isReadOnly={isReadOnly}
          loadRows={loadRows}
        />
      </div>
    )
  }
}

function wfModuleStatus (wfModule) {
  // TODO don't copy/paste from OutputPane.js
  if (wfModule.nClientRequests > 0) {
    // When we've just sent an HTTP request and not received a response,
    // mark ourselves "busy". This is great for when the user clicks "fetch"
    // and then is waiting for the server to set the status.
    //
    // The state stores server data separately than client data, so there's
    // no race when setting status and so if the "fetch" does nothing and the
    // server doesn't change wfModule.status, the client still resets its
    // perceived status.
    return 'busy'
  } else if (wfModule.is_busy) {
    return 'busy'
  } else if (!wfModule.output_status) {
    // placeholder? TODO verify this can actually happen
    return 'busy'
  } else {
    return wfModule.output_status
  }
}

function mapStateToProps (state) {
  const { workflow, wfModules, tabs } = state
  const tabSlug = workflow.tab_slugs[workflow.selected_tab_position]
  const tab = tabs[tabSlug]
  const wfModuleArray = tab.wf_module_ids.map(id => wfModules[String(id)])

  const wfModuleIndex = tab.selected_wf_module_position
  let wfModule = wfModuleArray[wfModuleIndex] || null
  let wfModuleBeforeError

  const status = wfModule ? wfModuleStatus(wfModule) : 'busy'

  if (wfModule === null && tab.wf_module_ids[wfModuleIndex]) {
    // We're pointing at a "placeholder" module: its id isn't in wfModules.
    // HACK: for now, we want OutputPane to render something different (it needs
    // to give TableSwitcher a "busy"-status WfModule).
    wfModule = {
      id: -1,
      html_output: false,
      status: 'busy',
      cached_render_result_delta_id: null,
      columns: null,
      nRows: null
    }
  }

  // If we're pointing at a module that output an error, we'll want to display
  // its _input_ (the previous module's output) to help the user fix things.
  if (status === 'error' && tab.selected_wf_module_position > 0) {
    const lastGood = wfModuleArray[wfModuleIndex - 1]
    wfModuleBeforeError = {
      id: lastGood.id,
      deltaId: lastGood.cached_render_result_delta_id,
      status: wfModuleStatus(lastGood),
      columns: lastGood.output_columns,
      nRows: lastGood.output_n_rows
    }
  }

  return {
    workflowId: workflow.id,
    wfModule: wfModule ? {
      id: wfModule.id,
      htmlOutput: wfModule.html_output,
      status,
      deltaId: wfModule.cached_render_result_delta_id,
      columns: wfModule.output_columns,
      nRows: wfModule.output_n_rows
    } : null,
    wfModuleBeforeError,
    isPublic: workflow.public,
    isReadOnly: workflow.read_only
  }
}

function mapDispatchToProps (dispatch) {
  return {
    loadRows: (wfModuleId, deltaId, startRow, endRow) => {
      return dispatch((_, __, api) => {
        return api.render(wfModuleId, startRow, endRow) // ignore deltaId -- for now
      })
    }
  }
}

export default connect(
  mapStateToProps,
  mapDispatchToProps
)(withI18n()(OutputPane))
