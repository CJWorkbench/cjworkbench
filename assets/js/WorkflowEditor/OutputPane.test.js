/* global describe, it, expect, jest */
import React from 'react'
import { OutputPane } from './OutputPane'
import OutputIframe from '../OutputIframe'
import DelayedTableSwitcher from '../table/DelayedTableSwitcher'
import { shallowWithI18n } from '../i18n/test-utils'

describe('OutputPane', () => {
  const wrapper = function (extraProps = {}) {
    return shallowWithI18n(
      <OutputPane
        loadRows={jest.fn()}
        workflowId={123}
        wfModule={{ id: 987, deltaId: 1, status: 'ok', htmlOutput: false }}
        isPublic={false}
        isReadOnly={false}
        {...extraProps}
      />
    )
  }

  it('matches snapshot', () => {
    const w = wrapper()
    expect(w).toMatchSnapshot()
  })

  it('renders a DelayedTableSwitcher', () => {
    const w = wrapper()
    expect(w.find(DelayedTableSwitcher)).toHaveLength(1)
  })

  it('renders when no module id', () => {
    const w = wrapper({ wfModule: null })
    expect(w).toMatchSnapshot()
    expect(w.find(DelayedTableSwitcher)).toHaveLength(1)
  })

  it('renders an iframe when htmlOutput', () => {
    const w = wrapper({ wfModule: { id: 1, deltaId: 2, htmlOutput: true, status: 'ok' } })
    expect(w.find(OutputIframe).prop('visible')).toBe(true)

    // When !htmlOutput, we just set visible=false but continue to display it.
    // That's because react-data-grid would have the wrong size otherwise.
    const w2 = wrapper({ wfModule: { id: 1, deltaId: 2, htmlOutput: false, status: 'ok' } })
    expect(w2.find(OutputIframe).prop('visible')).toBe(false)
  })

  it('renders different table than iframe when desired', () => {
    const w = wrapper({
      // even if before-error has htmlOutput, we won't display that one
      wfModuleBeforeError: { id: 1, deltaId: 2, status: 'ok', htmlOutput: true },
      wfModule: { id: 3, deltaId: 4, status: 'error', htmlOutput: true }
    })
    expect(w.find(DelayedTableSwitcher).prop('wfModuleId')).toEqual(1)
    expect(w.find(DelayedTableSwitcher).prop('deltaId')).toEqual(2)
    expect(w.text()).toMatch(/This was the data that led to an error./)
    expect(w.find(OutputIframe).prop('wfModuleId')).toEqual(3)
    expect(w.find(OutputIframe).prop('deltaId')).toEqual(4)
  })
})
