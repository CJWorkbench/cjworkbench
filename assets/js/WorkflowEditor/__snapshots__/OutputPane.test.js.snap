// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`OutputPane matches snapshot 1`] = `
<div
  className="outputpane module-ok"
>
  <Connect(WithI18n)
    deltaId={1}
    isPublic={false}
    key="iframe"
    visible={false}
    wfModuleId={987}
    workflowId={123}
  />
  <DelayedTableSwitcher
    deltaId={1}
    isReadOnly={false}
    key="table"
    loadRows={[MockFunction]}
    status="ok"
    wfModuleId={987}
  />
</div>
`;

exports[`OutputPane renders when no module id 1`] = `
<div
  className="outputpane module-unreachable"
>
  <Connect(WithI18n)
    deltaId={null}
    isPublic={false}
    key="iframe"
    visible={false}
    wfModuleId={null}
    workflowId={123}
  />
  <DelayedTableSwitcher
    columns={null}
    deltaId={null}
    isReadOnly={false}
    key="table"
    loadRows={[MockFunction]}
    nRows={null}
    status={null}
    wfModuleId={null}
  />
</div>
`;
